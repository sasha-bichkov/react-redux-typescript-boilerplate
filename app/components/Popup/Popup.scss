.Popup__overlay {
  position: fixed;
  top: 0;
  left: 0;
  z-index: -1;

  background-color: rgb(0 0 0 / .4);

  @include size(100%, 100%);
}

.Popup {
  position: absolute;
  top: 50%;
  left: 50%;
  display: flex;
  align-items: center;
  justify-content: center;

  min-width: 300px;

  transform: translate(-50%, -50%);
}

.Popup__content {
  display: flex;
  flex-direction: column;

  max-width: 260px;
  min-height: 150px;
  padding: 20px;
  border-radius: 12px;

  @include mq($from: tablet) {
    max-width: 700px;
  }
}

.Popup__content-decoration {
  background-color: white;
  box-shadow: rgb(0 0 0 / .5) 0 16px 70px;
}

.Popup__contentHeader {
  align-self: flex-end;
}

.Popup__contentBody {
  padding: 15px 0;
}

.Popup__clearButton {
  padding: 10px;
  border: none;
  background-color: transparent;
  color: inherit;
  font: inherit;

  cursor: pointer;
}

.Popup__closeButton {
  color: #000;
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 1;
  text-shadow: 0 1px 0 #fff;

  opacity: .5;
  cursor: pointer;
  transition-duration: .4s;

  &:hover,
  &:active {
    color: #000;
    text-decoration: none;

    opacity: .75;
  }

  &:focus {
    color: black;
    font-size: 2rem;

    opacity: 1;
  }
}

// used this library for animation https://animate.style/

.Popup__content-fadeIn {
  animation: .5s fade-in;
}

.Popup__content-backInDown {
  animation: .5s back-in-down;
}

@keyframes back-in-down {
  0% {
    opacity: .7;
    transform: translateY(-1200px) scale(.7);
  }

  80% {
    opacity: .7;
    transform: translateY(0) scale(.7);
  }

  100% {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes fade-in {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
