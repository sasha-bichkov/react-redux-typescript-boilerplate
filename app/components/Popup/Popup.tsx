import React, { FunctionComponent, ReactNode } from 'react'
import FocusLock from 'react-focus-lock'

import './Popup.scss'

export interface PropTypes  {
  readonly overlay: boolean
  readonly children?: ReactNode
  close(): void
}

const Popup: FunctionComponent<PropTypes> = ({
  overlay,
  close,
  children
}) => {
  return (
    <React.Fragment>
      <div className="Popup Popup-decoration">
        <FocusLock>
          <div className="Popup__content Popup__content-decoration Popup__content-backInDown">
            <div className="Popup__contentHeader">
              <button className="Popup__closeButton Popup__clearButton" onClick={close}>
                <span aria-hidden="true">Ã—</span>
              </button>
            </div>
            <div className="Popup__contentBody">
              {children}
            </div>
          </div>
        </FocusLock>
      </div>
      {overlay && (<div className="Popup__overlay" />)}
    </React.Fragment>
  )
}

export default Popup
