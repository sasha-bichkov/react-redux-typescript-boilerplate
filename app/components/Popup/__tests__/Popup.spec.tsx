import React from 'react'
import { mount } from 'enzyme'
import Popup from '../Popup'

describe('Popup', () => {
  it('calls button onclick callback at once if click button', () => {
    const mockCallback = jest.fn()
    const component = mount(
      <Popup
        overlay={true}
        close={() => console.log('click')}
      />
    )

    component.find('.Popup__closeButton').simulate('click')

    expect(mockCallback).toHaveBeenCalledTimes(0)
  })

  describe('if overlay is true', () => {
    it('renders background', () => {
      const component = mount(
        <Popup
          overlay
          close={() => console.log('click')}
        />
      )

      const result = component.find('.overlay')

      expect(result).toHaveLength(1)
    })
  })

  describe('if overlay is false', () => {
    it('does not render background', () => {
      const component = mount(
        <Popup
          overlay={false}
          close={() => console.log('click')}
        />
      )

      const result = component.find('.overlay')

      expect(result).toHaveLength(0)
    })
  })
})
