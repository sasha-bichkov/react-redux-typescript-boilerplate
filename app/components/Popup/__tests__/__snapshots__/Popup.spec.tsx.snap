// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should render Popup component 1`] = `
<Fragment>
  <div
    className="Popup Popup-decoration"
  >
    <ForwardRef(FocusLockUICombination)>
      <div
        className="Popup__content Popup__content-decoration backInDown"
      >
        <div
          className="Popup__contentHeader"
        >
          <button
            className="Popup__closeBtn clearBtn"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          className="Popup__contentBody"
        />
      </div>
    </ForwardRef(FocusLockUICombination)>
  </div>
  <div
    className="overlay"
  />
</Fragment>
`;
